tiled/readers/dataframe.py

```diff
@@ -45,11 +45,50 @@ def read_csv(cls, *args, metadata=None, **kwargs):
         >>> DataFrameAdapter.read_csv("myfiles.*.csv")
         >>> DataFrameAdapter.read_csv("s3://bucket/myfiles.*.csv")
         """
-        return cls(dask.dataframe.read_csv(*args, **kwargs), metadata=metadata)
+        return cls(lambda: dask.dataframe.read_csv(*args, **kwargs), metadata=metadata)
 
     @classmethod
-    def from_pandas(cls, *args, metadata=None, **kwargs):
-        return cls(dask.dataframe.from_pandas(*args, **kwargs), metadata=metadata)
+    def read_csv(cls, *args, metadata=None, **kwargs):
+        """
+        Read a CSV.
+
+        Internally, this uses dask.dataframe.read_csv.
+        It forward all parameters to that function. See
+        https://docs.dask.org/en/latest/dataframe-api.html#dask.dataframe.read_csv
+
+        Examples
+        --------
+
+        >>> DataFrameAdapter.read_csv("myfiles.*.csv")
+        >>> DataFrameAdapter.read_csv("s3://bucket/myfiles.*.csv")
+        """
+        cache = get_internal_cache()
+        ddf = dask.dataframe.read_csv(*args, **kwargs)
+        # A dask.dataframe does not know its byte size, so we use npartitions == 1
+        # as a proxy for "small enough to cache".
+        if (cache is not None) and ddf.npartitions == 1:
```

Comment:
> Why is the cache used in the read_csv classmethod constructor and not more deeply inside the instance? This makes it easy to make alternative constructors with different caching choices (e.g. different threshold, different cache key, ...).

This sounds sensible but you might end up with many similar constructors across adapters. Perhaps a util module could reduce the duplication somewhat? You could have a shared `CacheSettings` dataclass and some functions which hold the non-`read_csv`-specific logic in here. You would still have room to add adapter-specific caching arguments to constructors.

Reply:
Yeah, this looks way too complex to be replicated into every Adapter. I think you are right. In fact, we can leverage dask's [opportunistic caching](https://docs.dask.org/en/latest/caching.html) mechanism for every adapter that uses dask. (But we don't have to restrict the cache to _only_ work with dask.) Commit on the way....

Reply:
This is now much improved and simplified. Thanks for the feedback,  @callumforrester.